include/oneapi/dpl/internal/distributed_ranges_impl/detail/sycl_utils.hpp

```diff
@@ -0,0 +1,258 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <limits>
+
+#include <oneapi/dpl/internal/distributed_ranges_impl/detail/utils.hpp>
+
+#ifdef SYCL_LANGUAGE_VERSION
+
+#    include <sycl/sycl.hpp>
+
+namespace oneapi::dpl::experimental::dr::__detail
+{
+
+// With the ND-range workaround, the maximum kernel size is
+// `std::numeric_limits<std::int32_t>::max()` rounded down to
+// the nearest multiple of the block size.
+inline std::size_t
+max_kernel_size_(std::size_t block_size = 128)
+{
+    std::size_t max_kernel_size = std::numeric_limits<std::int32_t>::max();
+    return (max_kernel_size / block_size) * block_size;
+}
+
+// This is a workaround to avoid performance degradation
+// in DPC++ for odd range sizes.
```

Comment:
Is this something that needs to be discussed? (Is the performance degradation still observed / workaround still necessary?)

Can we contribute improvements to pattern_parallel_for / `oneapi::dpl::experimental::ranges::for_each`?